wallet: Refactor the classes in wallet/db.{cpp/h}

achow101 opened the pull request:
The data storage layer of the wallet is fairly complicated and not well encapsulated or modularized. This makes it difficult for us to introduce new data storage methods. So this PR refactors it such that there is a clear way to introduce other storage options. This is also intended to clean up the logic of the database layer so that it is easier to follow and reason about.

In the database, there are 3 classes: `BerkeleyBatch`, `BerkeleyDatabase`, and `BerkeleyEnvironment`. `BerkeleyDatabase` is typenamed to `WalletDatabase`. The goal is to introduce two abstract classes: `WalletDatabase` and `DatabaseBatch`. These will be implemented by `BerkeleyDatabase` and `BerkeleyBatch`. This abstract classes can be inherited by other storage methods. All of the database specific things should be hidden in those classes.

To achieve this, we move the database handling things in `BerkeleyBatch` and turn it purely into a data accessor for the `BerkeleyDatabase`. Particularly, various static functions that operated on a `BerkeleyDatabase` are changed to be member functions of `BerkeleyDatabase`. `Read`, `Write`, `Erase`, and `Exists` are refactored to have separate functions that can be overridden by other classes. Instead of `GetCursor` returning a BDB specific `Dbc` object, new `CreateCursor` and `CloseCursor` functions are added and the cursor is handled internally within the `BerkeleyBatch`. We are then able to introduce the `DatabaseBatch` abstract class.

Functionality is further moved out of `BerkeleyBatch` into `BerkeleyDatabase`. Notably the `Db` handle creation is moved from `BerkeleyBatch`'s constructor and to a new `Open` function in `BerkeleyDatabase`. `BerkeleyDatabase` will also handle closing with a new `Close` function instead of having this be part of `Flush`.

Additionally, the existence of `BerkeleyEnvironment` is hidden inside of `BerkeleyDatabase`. `mapFileUseCount` is removed. Instead `WalletDatabase` will track it's use count with `m_refcount`. This is incremented by newly introduced `AddRef` and `RemoveRef` functions. `m_refcount` is used to ensure that the database and environment are not closed while the database may still be in use.

Instead of each `BerkeleyEnvironment` tracking the fileids of the databases in that environment, a global map `g_fileids` is used. Since we were already going through every open fileid for the uniqeness check, it doesn't matter what `BerkeleyEnvironment` has a database with a particular fileid. All we care about is that there is a database in any of the environments that has the same fileid of the database that is currently being opened.

Lastly, `BerkeleyBatch, `BerkeleyDatabase`, and `BerkeleyEnvironment` are moved into new files `wallet/bdb.{cpp/h}`. `WalletDatabase` and `DatabaseBatch` are in `wallet/db.{cpp/h}`. This just gives better organization of the code so they aren't all mixed together.

***

I've started breaking this down into separate PRs. Some of the simpler stuff is moved up to the front so they can be merged first.

* [x] Mostly simple moveonly things: #19290
  * walletdb: Make SpliWalletFilePath non-static
  * walletdb: Add IsBDBWalletLoaded to look for BDB wallets specifically
  * walletdb: move IsWalletLoaded to walletdb.cpp
  * walletdb: moveonly: Move BerkeleyBatch Cursor and Txn funcs to cpp
  * walletdb: Move BDB specific things into bdb.{cpp/h}

These PRs are independent of each other and can be merged out of order after #19290 is merged
* [x] Refactor Read, Write, Erase, Exists: #19292
  * walletdb: refactor Read, Write, Erase, and Exists into non-template func
* [x] Handle cursor internally: #19308
  * walletdb: Handle cursor internally
* [x] Make Create, CreateMock, and CreateDummy standalone: #19310
  * Add Create*WalletDatabase functions
  * scripted-diff: Replace WalletDatabase::Create* with CreateWalletDatabase
  * Remove WalletDatabase::Create, CreateMock, and CreateDummy
* [x] Move BerkeleyBatch static functions: #19324
  * walletdb: Combine VerifyDatabaseFile and VerifyEnvironment
  * walletdb: Move PeriodicFlush into WalletDatabase
  * walletdb: Move Rewrite into BerkeleyDatabase

These PRs require the previous listed to be merged first and need to be merged in order
* [x] Introduce `DatabaseBatch`: #19325
  * walletdb: Refactor DatabaseBatch abstract class from BerkeleyBatch
  * walletdb: Add MakeBatch function to BerkeleyDatabase and use it
* [x] Introduce `WalletDatabase`: #19334
  * walletdb: Move BerkeleyDatabase::Flush(true) to Close()
  * walletdb: Introduce AddRef and RemoveRef functions
  * walletdb: Add BerkeleyDatabase::Open dummy function
  * walletdb: Introduce WalletDatabase abstract class
* [ ] Cleanup the separation: #19335
  * walletdb: track database file use as m_refcount within BerkeleyDatabase
  * walletdb: Move Db->open to BerkeleyDatabase::Open
  * walletdb: Use a global g_fileids instead of m_fileids for each env
  * walletdb: Remove BerkeleyBatch friend class from BerkeleyDatabase

I will continue to break this down as things get merged.

achow101 committed:
walletdb: Move Db->open to BerkeleyDatabase::Open

Instead of opening the Db handle in BerkeleyBatch, make BerkeleyDatabase
do that.

achow101 committed:
walletdb: Use a global g_fileids instead of m_fileids for each env

File ids were basically global anyways, it doesn't make sense to have
them be environment specific if we're going to loop through all the
environments anyways.

achow101 committed:
walletdb: track database file use as m_refcount within BerkeleyDatabase

Instead of having BerkeleyEnvironment track the file use count, make
BerkeleyDatabase do it itself.

meshcollider commented:
Concept ACK

ryanofsky commented:
Strong concept ACK for the effort and end result, but I'm scared it looks like the intermediate commits here are moving and changing code at the same time instead of just leaving code where it is or moving it in trivial to review `--color-moved` MOVEONLY commits. This was previously a hurdle for me in https://github.com/bitcoin/bitcoin/pull/16341#issuecomment-541330425, but maybe the problem is less severe here because it looks like the moves are happening in single commits, not in harder to reconcile add and remove commits.

I would definitely appreciate it (can't speak for other reviewers) if this PR could be updated to use MOVEONLY commits (multiple small ones throughout or a big one at the beginning or end), but I'll give this a try and maybe the review will be easier than I'm expecting 👍

achow101 committed:
walletdb: Handle cursor internally

Instead of returning a Dbc (BDB cursor object) and having the caller
deal with the cursor, make BerkeleyBatch handle the cursor internally.

This prepares BerkeleyBatch to work with other database systems as Dbc
objects are BDB specific.

achow101 commented:
> but I'm scared it looks like the intermediate commits here are moving and changing code at the same time instead of just leaving code where it is or moving it in trivial to review

I only found one commit which was moving and changing behavior at the same time. This was `wallet: Move cursor functions to WalletDatabase and handle internally` which I have now split into `walletdb: Handle cursor internally` (changes behavior) and `walletdb: Move cursor functions to WalletDatabase` (move).

bvbfan commented:
The patch changes ref counting of db files. Now `Acquire` is called only in `Rewrite` since previous batch do it. Now ref count is always 0 (except you call Rewrite) which makes whole read/write racy.

achow101 commented:
> The patch changes ref counting of db files. Now `Acquire` is called only in `Rewrite` since previous batch do it. Now ref count is always 0 (except you call Rewrite) which makes whole read/write racy.

`Acquire` is called by `WalletBatch`'s constructor so the ref count is still incremented every time the database is being used. And `Release` is called by `WalletBatch`'s destructor.

Empact commented on src/wallet/wallettool.cpp:
```diff
@@ -103,6 +105,207 @@ static void WalletShowInfo(CWallet* wallet_instance)
+/* End of key/value data */
+static const char *DATA_END = "DATA=END";
+
+static bool SalvageWallet(fs::path path)
```
nit: `const fs::path&=

achow101 replied:
This commit is part of #18918 so comments should be left there.

Empact commented on src/wallet/init.cpp:
```diff
@@ -54,7 +54,6 @@ void WalletInit::AddWalletOptions() const
     gArgs.AddArg("-paytxfee=<amt>", strprintf("Fee (in %s/kB) to add to transactions you send (default: %s)",
                                                             CURRENCY_UNIT, FormatMoney(CFeeRate{DEFAULT_PAY_TX_FEE}.GetFeePerK())), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     gArgs.AddArg("-rescan", "Rescan the block chain for missing wallet transactions on startup", ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
-    gArgs.AddArg("-salvagewallet", "Attempt to recover private keys from a corrupt wallet on startup", ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
IMO "Attempt to recover private keys from a corrupt wallet" explains salvage more clearly than "Perform salvage operations on a wallet file"

achow101 replied:
This commit is part of #18918 so comments should be left there.

Empact commented:
@achow101 re @ryanofsky's comment, I set out to expand and clarify https://github.com/bitcoin/bitcoin/pull/18971/commits/67e30b2b42b4d710739e6fe303cbca69be68032c into several simpler move-only commits. 

You can use `git diff 67e30b2b42b4d710739e6fe303cbca69be68032c 125796d8e60b8e740857e427cbaad4aa23f74b52` to compare the results, as there are some changes I've left out, and slight difference, e.g. I took the approach of exposing `WalletBatch::ReadKeyValue`, but they don't affect the later commits.

https://github.com/achow101/bitcoin/compare/refactor-storage...Empact:refactor-storage-moveonly

achow101 commented:
@Empact Thanks for doing that! Unfortunately I worked out my own way of doing the separation before I saw your comment. My version was pushed to #18918.

bvbfan commented:
> `Acquire` is called by `WalletBatch`'s constructor so the ref count is still incremented every time the database is being used. And `Release` is called by `WalletBatch`'s destructor.

Got'cha, sorry about that, i've not notice it. I still have concern about touching db env in multi threading. `BerkeleyDatabase` should guard every env usage, for example in contructor with shared env, the easiest test is to call `CreateWalletFromFile` from distinct threads with same db location, it will crash at 
`auto inserted = this->env->m_databases.emplace(strFile, std::ref(*this));`
before actual assert in next line.

Sjors commented:
Concept ACK

jonatack commented on src/wallet/walletdb.h:
```diff
@@ -287,11 +290,47 @@ class WalletBatch
+
+//! Unserialize a give Key-Value pair and load it into the wallet and wallet scan state
+bool ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+             CWalletScanState &wss, std::string& strType, std::string& strErr) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet);
```
When building with clang --enable-debug -Werror I'm seeing the following compiler error:

```cpp
./wallet/walletdb.h:322:112: error: member access into incomplete type 'CWallet'
             CWalletScanState &wss, std::string& strType, std::string& strErr) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet);
                                                                                                               ^
./wallet/ismine.h:14:7: note: forward declaration of 'CWallet'
class CWallet;
      ^
1 error generated.
```

jonatack replied:
(It may not be directly related to this PR; saw a similar build error in another wallet PR today.)

jonatack reviewed (commented):
Concept ACK. A quick scan of each commit looks reasonable. I was looking at this area of the codebase the other day with an eye to fixing the TODOs in tool_wallet.py and these changes look like the code would be better organised and easier to modify and reason about.

achow101 commented:
Not sure what's causing the travis failure. It seems to be caused for a non-existent fileid when `get_fileid` is called, but I can't figure out why that would be the case nor can I replicate it locally.

bvbfan commented:
> Not sure what's causing the travis failure. It seems to be caused for a non-existent fileid when `get_fileid` is called, but I can't figure out why that would be the case nor can I replicate it locally.

`get_fileid` is not called in mock db (some older bdb version in Travis?)

achow101 commented:
> `get_fileid` should not be called in mock db

Ah, that's probably it. Thanks. Odd that I can't get the error locally though.

> (some older bdb version in Travis?)

It's actually newer. Travis typically uses BDB provided by Ubuntu which is 5.3. I have 4.8 installed on my system.

bvbfan commented:
> It's actually newer. Travis typically uses BDB provided by Ubuntu which is 5.3. I have 4.8 installed on my system.

I can't reproduce it either, bdb 5.3.28.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3680,17 +3680,7 @@ bool CWallet::Verify(interfaces::Chain& chain, const WalletLocation& location, b
-        if (!WalletBatch::VerifyEnvironment(wallet_path, error_string)) {
-            return false;
-        }
-    } catch (const fs::filesystem_error& e) {
```
In commit "walletdb: Combine VerifyDatabaseFile and VerifyEnvironment" (289aa21c6a6c11ce172a92f878a3ede6652fc2e4)

Here and below filesystem_error is no longer caught. Would suggest adding it back to preserve error handling behavior (I could be missing something but it looks like BerkeleyEnvironment::Open could still throw this error)

achow101 replied:
Added it back.

ryanofsky commented on src/wallet/db.cpp:
```diff
@@ -351,13 +351,26 @@ void BerkeleyEnvironment::CheckpointLSN(const std::string& strFile)
     fReadOnly = (!strchr(pszMode, '+') && !strchr(pszMode, 'w'));
     fFlushOnClose = fFlushOnCloseIn;
     env = database.env.get();
-    if (database.IsDummy()) {
```
In commit "walletdb: Move Db->open to BerkeleyDatabase::Open" (d0ccd47c355319ec9709c52cdcdf0be78a26ac25)

Note: seems like there two small changes to behavior here when IsDummy is true: CLIENT_VERSION is written and strFile member is set, but both changes seem good.

ryanofsky reviewed (commented):
Started review (will update list below with progress). Nice cleanups so far

- [X] e9dff450d65983644f5c60e1c6aa1f65ade5619e walletdb: Combine VerifyDatabaseFile and VerifyEnvironment (1/20)
- [X] e1c46d4de66d57a703a9aa1b4d1b2bd65e182445 walletdb: Move PeriodicFlush into WalletDatabase (2/20)
- [X] 3cdb7fd9d1e92f07c04c325c9b585ff736162198 walletdb: Move Db->open to BerkeleyDatabase::Open (3/20)
- [X] b6a6e0292c226f67b9ee177f9732a936aa7a0d7e walletdb: Use a global g_fileids instead of m_fileids for each env (4/20)
- [X] 8726d373e165087789933eb03383dd7feee31d3f walletdb: Add BerkeleyDatabase::Release (5/20)
- [X] 2c663e453104b2f49849d8bc80727e98abda8055 walletdb: Move log consolidation into BerkeleyEnvironment::Close (6/20)
- [X] f3b553f36e83c44a4edb51710ae673d6b9b3e573 walletdb: Change BerkeleyDatabase::Flush to Close (7/20)
- [X] 9603e233c9c4049a267282615b5e4f31600656e6 walletdb: track database file use as m_refcount within BerkeleyDatabase (8/20)
- [X] 17009d1e73ce894f9f5d5441c472e87efaee6163 walletdb: Replace BerkeleyDatabase::Flush with BerkeleyBatch::Flush (9/20)
- [X] 2711caa2a2d4f04b848c25b614b353a1a08351b0 walletdb: Handle cursor internally (10/20)
- [X] 7d8b1b42702357d5c25e5825dfc47326bad0568f walletdb: Move cursor functions to WalletDatabase (11/20)
- [X] 75f3532a70abf4331aba212f979a2e8be02d191c walletdb: Move Rewrite into BerkeleyDatabase (12/20)
- [X] e1ebc69a5e6628cadf480dfd6d787a6c9f631b86 walletdb: Move Txn* functions into WalletDatabase (13/20)
- [X] 7f6608ba7ac54935ddebec9b49817fc4a26e749e walletdb: Move Read, Write, Erase, and Exists into BerkeleyDatabase (14/20)
- [X] 64a5a29d692378df443ae599b81c1f4547ec3d03 walletdb: Remove BerkeleyBatch and have WalletBatch use WalletDatabase (15/20)
- [X] aac3483e9e0423b3414cfaf84f37041a200d773c walletdb: refactor Read, Write, Erase, and Exists into non-template func (16/20)
- [X] 6091f36d30f1fd9feec4bef6677ce514b5d20b2c walletdb: Move BerkeleyDatabase::Create, CreateDummy, and CreateMock out (17/20)
- [X] 4c6c9080d48ec5810e7d36348f7c730a586e9fb5 walletdb: Make WalletDatabase abstract class (18/20)
- [X] d858c9b23699edaa27166071fc0c88f55911fea8 wallet: Add IsBDBWalletLoaded to look for BDB wallets specifically (19/20)
- [X] b4c2765fd19e5dd8c823febd65ad5e62a58c8260 Move BDB specific things into bdb.{cpp/h} (20/20)

ryanofsky commented on src/wallet/db.cpp:
```diff
@@ -531,11 +529,10 @@ bool BerkeleyBatch::Rewrite(BerkeleyDatabase& database, const char* pszSkip)
                 // Flush log data to the dat file
                 env->CloseDb(strFile);
                 env->CheckpointLSN(strFile);
-                env->mapFileUseCount.erase(strFile);
```
In commit "walletdb: track database file use as m_refcount within BerkeleyDatabase" (9cfb631f808aa6ccb10ab31d4b8325bd940a8916)

Note: seems this commit drops distinction between use count being 0 and use count not existing (no map entry). I think 0 use count meant the database wasn't in use but wasn't flushed yet, and no use count meant it was flushed. I think the last piece of code which used this distinction was the log_archive code removed in the previous commit 1dbc24d27d4b54b45c6b49ba41b4f64d0093dd71

ryanofsky reviewed (commented):
Updated https://github.com/bitcoin/bitcoin/pull/18971#pullrequestreview-419340191

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -3682,15 +3682,11 @@ bool CWallet::Verify(interfaces::Chain& chain, const WalletLocation& location, b
         return false;
     }
-
-    return WalletBatch::VerifyDatabaseFile(wallet_path, error_string);
```
nit: assert as unreachable?

achow101 replied:
Done

Empact commented on src/wallet/wallettool.cpp:
```diff
@@ -141,10 +142,14 @@ bool ExecuteWalletToolFunc(const std::string& command, const std::string& name)
         }
         bilingual_str error;
-        if (!WalletBatch::VerifyEnvironment(path, error)) {
+        std::vector<bilingual_str> warnings;
```
nit: this is unused

achow101 replied:
Removed

achow101 commented on src/wallet/db.h:
```diff
@@ -143,9 +143,12 @@ class BerkeleyDatabase
+    /** Open the database if it is not already opened. Increments mapFileUseCount */
     void Open(const char* mode);
 
+    /** Indicate that database user has stopped using the database. Decrement mapFileUseCount */
```
I think you are confusing `BerkeleyBatch` and `BerkeleyDatabase`. `Release` is called by `BerkeleyBatch::Close`, not `BerkeleyDatabase::Close`.

In a later commit, a corresponding `Acquire` function is added. In this commit, the thing that `Acquire` does is part of `Open`. The idea with the refcount is thusly: when something tries to use the database, it increments the refcount. When it is done, it decrements the refcount. The goal is to prevent the database from closing while something is still using the database. When `WalletBatch` is initialized, it calls `BerkeleyDatabase::Open` via `BerkeleyBatch`'s constructor. When `WalletBatch` is done and no longer being used, it calls `BerkeleyBatch::Close` which calls `BerkeleyDatabase::Release`. Then both the `BerkeleyBatch` and `WalletBatch` are destructed. The `BerkeleyDatabase` remains open so that future uses of the database don't need to reopen it and for other concurrent `WalletBatch`s to do their work.

So the flow with `BerkeleyDatabase` is `Open`, changes, `Release`, `Close`.

achow101 replied:
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -350,6 +355,18 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bo
+        LOCK(cs_db);
+        size_t erased = env->m_databases.erase(strFile);
+        assert(erased == 1);
+        g_dbenvs.erase(env->Directory().string());
```
Not entirely sure, but this is moved code.

This was originally called during `BerkeleyDatabase::Flush(true)`, i.e. when the database and environment are being closed. So I moved it into the destructor to go along with the other environment closing changes.

achow101 replied:
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -619,17 +637,7 @@ void BerkeleyEnvironment::Flush(bool fShutdown)
                 mi++;
         }
-        LogPrint(BCLog::WALLETDB, "BerkeleyEnvironment::Flush: Flush(%s)%s took %15dms\n", fShutdown ? "true" : "false", fDbEnvInit ? "" : " database not started", GetTimeMillis() - nStart);
-        if (fShutdown) {
```
> Q: If there are multiple databases open in the same environment, previously logs would be consolidated and removed if any database was closed and no other database was writing, but now it will not consolidate logs until the last database in the environment is closed?

I don't think that's what happened previously either. `Flush(true)` seems to have been only called during a bitcoind shutdown, not during typical wallet unloading. So previously, even after a wallet was unloaded, it's log files were never consolidated and removed. So this behavior ends up being retained. I don't think there is a behavior change here, although the it is difficult to work out exactly when the database and environment are actually being flushed and closed.

achow101 replied:
Also split this commit as you suggested.

achow101 replied:
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -27,22 +27,7 @@ namespace {
-        }
-    }
-}
+std::unordered_map<std::string, WalletDatabaseFileId> g_fileids;
```
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -405,24 +390,34 @@ void BerkeleyDatabase::Open(const char* pszMode)
+                            HexStr(std::begin(item.second.value), std::end(item.second.value)), item.first));
+                    }
+                }
+                g_fileids[m_file_path] = fileid;
```
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -720,22 +728,16 @@ bool BerkeleyDatabase::Backup(const std::string& strDest) const
-            g_fileids.erase(m_file_path);
-        }
+        env->Flush();
+        // TODO: To avoid g_dbenvs.erase erasing the environment prematurely after the
```
Removed it

achow101 commented on src/wallet/db.cpp:
```diff
@@ -652,30 +646,23 @@ bool BerkeleyDatabase::PeriodicFlush()
-                LogPrint(BCLog::WALLETDB, "Flushed %s %dms\n", strFile, GetTimeMillis() - nStart);
-                ret = true;
-            }
+            m_refcount = 0;
```
Removed

Sjors commented on src/wallet/db.cpp:
```diff
@@ -349,13 +349,26 @@ void BerkeleyEnvironment::CheckpointLSN(const std::string& strFile)
+
+void BerkeleyDatabase::Open(const char* pszMode)
+{
+    m_read_only = (!strchr(pszMode, '+') && !strchr(pszMode, 'w'));
```
In 3cdb7fd9d1e92f07c04c325c9b585ff736162198: it's a bit odd that you're introducing `m_read_only` here because it's not read anywhere yet. 64a5a29d692378df443ae599b81c1f4547ec3d03 seems a better place to introduce it.

achow101 replied:
I'd like to keep it here instead of removing and re-introducing this line.

Sjors commented on src/wallet/db.cpp:
```diff
@@ -377,19 +389,19 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bo
-                            fMockDb ? nullptr : strFilename.c_str(),  // Filename
-                            fMockDb ? strFilename.c_str() : "main",   // Logical db name
+                            fMockDb ? nullptr : strFile.c_str(),  // Filename
+                            fMockDb ? strFile.c_str() : "main",   // Logical db name
```
3cdb7fd9d1e92f07c04c325c9b585ff736162198 nit: rename breaks comment indent (it's still broken in `bdb.cpp` after the move)

achow101 replied:
Fixed

Sjors commented on src/wallet/db.cpp:
```diff
@@ -349,13 +349,26 @@ void BerkeleyEnvironment::CheckpointLSN(const std::string& strFile)
-    if (database.IsDummy()) {
+    pdb = database.m_db.get();
+    strFile = database.strFile;
+    if (strchr(pszMode, 'c') != nullptr && !Exists(std::string("version"))) {
```
3cdb7fd9d1e92f07c04c325c9b585ff736162198  nit: the `fCreate` helper variable makes this more readable. It's introduced in a later commit, but might as well do it here.

achow101 replied:
Done

Sjors reviewed (commented):
Reviewed the first couple of commits up to 3cdb7fd9d1e92f07c04c325c9b585ff736162198, will resume later.

ryanofsky commented on src/wallet/db.cpp:
```diff
@@ -755,9 +740,11 @@ std::string BerkeleyDatabaseVersion()
-        LOCK(cs_db);
-        --env->mapFileUseCount[strFile];
-    }
+    m_refcount--;
```
In commit "walletdb: track database file use as m_refcount within BerkeleyDatabase" (9603e233c9c4049a267282615b5e4f31600656e6)

Even though `m_refcount` is atomic you still have to lock cs_db while updating it, otherwise the `m_db_in_use.notify_all()` call may result in a lost wakeup and infinite hang if the `m_db_in_use.wait()` lambda happens to run at the same time `m_refcount` is decremented and return `false` (see https://stackoverflow.com/a/41875321 or https://github.com/isocpp/CppCoreGuidelines/issues/554).

achow101 replied:
Done

ryanofsky commented on src/wallet/db.cpp:
```diff
@@ -550,17 +550,17 @@ bool BerkeleyBatch::Rewrite(BerkeleyDatabase& database, const char* pszSkip)
 
-                    Dbc* pcursor = db.GetCursor();
-                    if (pcursor)
+                    if (db.CreateCursor())
```
In commit "walletdb: Handle cursor internally" (2711caa2a2d4f04b848c25b614b353a1a08351b0)

Behavior isn't changing in this commit, but error handling here is broken, right? This should be returning failure if creating a cursor doesn't succeed, not just pretending it succeeded and not reading anything.

achow101 replied:
Yes, I think error handling is broken here. I think the intent was to ignore an error and just keep trying until it succeeds. But there's nothing to actually have the outer while loop keep going.

ryanofsky commented on src/wallet/db.h:
```diff
@@ -332,27 +333,32 @@ class BerkeleyBatch
-        Dbc* pcursor = nullptr;
-        int ret = pdb->cursor(nullptr, &pcursor, 0);
+            return false;
+        m_cursor = nullptr;
```
In commit "walletdb: Handle cursor internally" (2711caa2a2d4f04b848c25b614b353a1a08351b0)

I think this needs to either return false if m_cursor is not null or assert that it is null, otherwise calling this at the wrong time will silently leak memory

achow101 replied:
Done

ryanofsky reviewed (commented):
Updated https://github.com/bitcoin/bitcoin/pull/18971#pullrequestreview-419340191

ryanofsky reviewed (approved):
Code review almost-ACK b4c2765fd19e5dd8c823febd65ad5e62a58c8260 (just various individual comments should be addressed).

This is a nice code improvement that enables new features (#19077, #19137, #19102)

My biggest comment in terms of scope is about sharing cursor and transaction objects. Sharing these seems fragile if not buggy (https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437330343). It seems like it would also shrink down the PR to just make the Batch cursor & transaction & read/write/erase methods virtual instead of trying to cram them into the already complicated Database class. Database class would just need a virtual `unique_ptr<Batch> Database::MakeBatch()` method returning the right Batch subclass for the database subclass.

I'm also surprised how the moveonly commits in this PR come after the change commits, rather than before. I'd be curious if you have a good workflow for updates because it seems like every update to one of the early change commits would require repeatedly making the same updates in the later move commits.

Anyway these are all fairly minor comments. The changes here are great and definitely much better than everything that was here before.

ryanofsky commented on src/wallet/db.cpp:
```diff
@@ -739,6 +747,10 @@ std::string BerkeleyDatabaseVersion()
 void BerkeleyDatabase::Release()
 {
+    if (m_active_txn) {
+       m_active_txn->abort();
```
In commit "walletdb: Move Txn* functions into WalletDatabase" ([e1ebc69](https://github.com/bitcoin/bitcoin/commit/e1ebc69a5e6628cadf480dfd6d787a6c9f631b86))

re: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437330343

> It's not coherent to have a shared reference count but only allow a single cursor and a single transaction at a time, and have different batches abort each others transactions and step on each others cursors.

I should be more specific about how I think this could be resolved. I think there are two ways:

1. Add a bunch of asserts to ensure the new shared cursors and transactions aren't misused. `assert(!m_active_txn)` in TxnBegin and Close and Release, `assert(m_active_txn)` in TxnCommit, TxnAbort, Read, Write, etc. `assert(!m_cursor)` in CreateCursor, etc.

2. Don't add shared cursors and transactions. Keep the methods where they are currently in `BerkeleyBatch` instead of moving them to `WalletDatabase`. Just make them them virtual overrides and have `BerkeleyBatch` inherit from a `DatabaseBatch` class. Give `WalletDatabase` a virtual `MakeBatch` method returning `unique_ptr<DatabaseBatch>` and have `WalletBatch` call it during construction and use the pointer as needed. You have two classes with virtual methods instead of one class with virtual methods, but less code moves around and the classes have clear responsibilities.

Another variants of the second approach are also possible like giving WalletBatch itself virtual methods, but the basic idea's the same.

achow101 replied:
Transactions are only used during encryption. Cursors are only used by loading, rewriting, and zapwallettx.

achow101 replied:
I've decided to go with option 1 and added asserts to `CreateCursor` and `TxnBegin`.

I don't think it makes much sense to support multiple transactions and cursors because of our extremely limited usage of them. Transactions are only used when encrypting a wallet. Cursors are used in scenarios where there is only a single thread accessing the database already. These are in loading (both load and zapwallettx) and rewrite, which is encryption. So there won't be concurrent cursors nor concurrent transactions. Even with multiple parallel RPCs, we use locking during encryption to enforce single access, and during loading, the wallet isn't setup yet to even have multiple accesses.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437707266

> Additionally, in SQLite, there aren't such a thing as multiple transactions, so having a separate `Batch/ class would make that more difficult.

RIGHT. :sweat_smile: Finally I understand why you are doing this. I thought you had just gone insane wanting to cram the database and batch classes together and make all these extra changes. But I remember using sqlite in python it doesn't support more than one transaction per connection.

Even so, I don't think this is the best choice or even the easier choice, because it is pushing a problem from the database layer up to the application and RPC layers. Now if there are two RPC calls happening in different threads that both want to open a transaction or a cursor, they have to coordinate with each other to see who is allowed to go first, or risk crashing the process.

Also, the C++ code is more complicated than it needs to be because instead of cursors and transactions being owned by batches and getting properly cleaned up when batches are cleaned up, now they can outlive batches, and error handling and cleanup that should be triggered when the batch goes out of scope now gets lost or delayed until a later point, making debugging more difficult. Also the nice layer separation where Batch objects are responsible for data access, and Database objects are responsible for management is lost, and more code has to move, and everything is more complicated.

If you like the current organization, then please ignore my suggestion, but here is what I would do:

- For BDB, put ownership of transaction and cursor objects back in the batch objects where they are happy and belong.

- For Sqlite, just have `Mutex m_cursor_mutex, m_transaction_mutex` members in the Database instance. Lock m_cursor when a batch object needs a cursor, and unlock when it's done. Similarly for m_transaction. That way if two RPC calls happen, then there aren't errors or aborts, one call just waits for the other.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r439666055

> Now if there are two RPC calls happening in different threads that both want to open a transaction or a cursor, they have to coordinate with each other to see who is allowed to go first, or risk crashing the process.

I forgot about cs_wallet when writing this. I think with cs_wallet, two batch objects for the same database will probably not even exist at the same time, so asserts only allowing one cursor or transaction should never trigger, and there should be no need for the `m_transaction_mutex` suggestion. Something like `bool m_transaction_started` and `assert(!m_transaction_started)` for sqlite would be fine. And there wouldn't be a need for `m_cursor_mutex` in any case since there's no problem with multiple cursors, just multiple transactions.

I still think it is better to keep different batch and connection objects for all the other reasons (limiting lifetime of cursor and txn objects, reporting errors earlier, smaller diff, smaller classes, separation of data access methods from database management methods), but with `cs_wallet` the multiple thread RPC request thing should be not be a real issue.

achow101 replied:
There does happen to be a few times where there are multiple `BerkeleyBatch` objects are active, although I don't know where this is happening. You can check this by adding some print lines at the place where `mapFileUseCount` is being updated.

I will probably rework this PR to keep `BerkeleyBatch` and make the move commits at the beginning so this isn't so difficult to update.

achow101 commented on src/wallet/db.h:
```diff
@@ -262,7 +211,121 @@ class BerkeleyDatabase
+    ~BerkeleyDatabase();
+
+    /** Database pointer. This is initialized lazily and reset during flushes, so it can be null. */
+    std::unique_ptr<Db> m_db;
```
BerkeleyEnvironment needs this.

achow101 commented on src/wallet/db.cpp:
```diff
@@ -25,24 +26,39 @@ namespace {
-    int ret = db.get_mpf()->get_fileid(fileid.value);
-    if (ret != 0) {
-        throw std::runtime_error(strprintf("BerkeleyBatch: Can't open database %s (get_fileid failed with %d)", filename, ret));
+static void AddUniqueFileId(Db& db, const std::string& path, const std::string& filename)
```
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -25,24 +26,39 @@ namespace {
+    LOCK(cs_db);
+    // Check that the BDB file id has not already been loaded in any BDB environment
+    // to avoid BDB data consistency bugs that happen when different data
+    // files in the same environment have the same fileid. All BDB environments are
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -430,7 +430,12 @@ bool CWallet::HasWalletSpend(const uint256& txid) const
+    if (database) {
+        database->Close();
+        if (shutdown) {
+            database.reset();
```
Removed.

achow101 commented on src/wallet/db.cpp:
```diff
@@ -740,7 +742,8 @@ void BerkeleyDatabase::Flush(bool shutdown)
             // environments can be shut down after databases.
-            env->m_fileids.erase(strFile);
+            LOCK(cs_db);
+            g_fileids.erase(m_file_path);
```
This can't be moved into CloseDb. However I have moved it into `BerkeleyEnvironment::Close`.

achow101 commented on src/wallet/db.cpp:
```diff
@@ -384,6 +384,18 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bo
+    if (env) {
+        LOCK(cs_db);
+        size_t erased = env->m_databases.erase(strFile);
+        assert(erased == 1);
```
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -371,6 +369,7 @@ void BerkeleyEnvironment::CheckpointLSN(const std::string& strFile)
 BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bool fFlushOnCloseIn) : pdb(nullptr), activeTxn(nullptr), m_database(database)
 {
     database.Open(pszMode);
+    database.Acquire();
```
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -536,7 +534,8 @@ bool BerkeleyBatch::Rewrite(BerkeleyDatabase& database, const char* pszSkip)
                 { // surround usage of db with extra {}
-                    BerkeleyBatch db(database, "r");
+                    Open("r");
+                    Acquire();
```
Done

achow101 commented on src/wallet/db.cpp:
```diff
@@ -536,7 +534,8 @@ bool BerkeleyBatch::Rewrite(BerkeleyDatabase& database, const char* pszSkip)
                 LogPrintf("BerkeleyBatch::Rewrite: Rewriting %s...\n", strFile);
                 std::string strFileRes = strFile + ".rewrite";
                 { // surround usage of db with extra {}
-                    BerkeleyBatch db(database, "r");
```
Done

achow101 commented on src/wallet/db.h:
```diff
@@ -124,6 +124,11 @@ class BerkeleyDatabase
+    bool DBRead(CDataStream& key, CDataStream& value) const;
+    bool DBWrite(CDataStream& key, CDataStream& value, bool overwrite=true) const;
+    bool DBErase(CDataStream& key) const;
+    bool DBExists(CDataStream& key) const;
```
Renamed and removed `const`

achow101 commented on src/wallet/walletdb.h:
```diff
@@ -182,35 +182,47 @@ class WalletBatch
         m_database(database)
     {
+        database.Open(pszMode);
+        database.Acquire();
```
Done

achow101 commented on src/wallet/db.h:
```diff
@@ -13,87 +13,24 @@
-#include <db_cxx.h>
 #if defined(__GNUC__) && !defined(__clang__)
 #pragma GCC diagnostic pop
 #endif
```
Removed

achow101 commented:
> My biggest comment in terms of scope is about sharing cursor and transaction objects. Sharing these seems fragile if not buggy ([#18971 (comment)](https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437330343)). It seems like it would also shrink down the PR to just make the Batch cursor & transaction & read/write/erase methods virtual instead of trying to cram them into the already complicated Database class. Database class would just need a virtual `unique_ptr<Batch> Database::MakeBatch()` method returning the right Batch subclass for the database subclass.

I'll look into it. However I'm wary of having parallel inheritances like this.

> I'd be curious if you have a good workflow for updates because it seems like every update to one of the early change commits would require repeatedly making the same updates in the later move commits.

No and it's very painful.

ryanofsky reviewed (approved):
Code review ACK 47dc8d6aff04e4f4c3ed0b4eb28fca0596d020e6. Looks very good. I suggested changes but they aren't too important and could all be followups.

> > I'd be curious if you have a good workflow for updates because it seems like every update to one of the early change commits would require repeatedly making the same updates in the later move commits.
> 
> No and it's very painful.

Sorry for that. My experience is it's less painful with a one or two big move commits up front (which can also be merged earlier) and the refactoring changes after the moves. But just having the moves separate at all has really helped me review this PR, so I definitely appreciate the effort!

achow101 committed:
walletdb: Combine VerifyDatabaseFile and VerifyEnvironment

Combine these two functions into a single Verify function that is a
member of WalletDatabase. Additionally, these are no longer static.

achow101 committed:
walletdb: Move PeriodicFlush into WalletDatabase

Make PeriodicFlush a non-static member of WalletDatabase instead of
WalletBatch.

achow101 commented on src/wallet/db.cpp:
```diff
@@ -126,6 +145,8 @@ void BerkeleyEnvironment::Close()
             database.m_db->close(0);
             database.m_db.reset();
+            LOCK(cs_db);
+            g_fileids.erase(database.m_file_path);
```
Do you have a specific scenario where this could happen? This change seems to be causing a test failure.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r440386812

> Do you have a specific scenario where this could happen? This change seems to be causing a test failure.

Marking resolved, followed up in a new thread: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r441652829

achow101 committed:
walletdb: Move Rewrite into BerkeleyDatabase

Make Rewrite actually a member of BerkeleyDatabase instead of a static
function in BerkeleyBatch

achow101 committed:
walletdb: Remove BerkeleyBatch friend class from BerkeleyDatabase

achow101 committed:
walletdb: Introduce WalletDatabase abstract class

Make WalletDatabase actually an abstract class and not just a typedef
for BerkeleyDatabase. Have BerkeleyDatabase inherit this class.

achow101 committed:
walletdb: Add MakeBatch function to BerkeleyDatabase and use it

Instead of having WalletBatch construct the BerkeleyBatch, have
BerkeleyDatabase do it and return a std::unique_ptr<BerkeleyBatch>

achow101 committed:
walletdb: Move BerkeleyDatabase::Flush(true) to Close()

Instead of having Flush optionally shutdown the database and
environment, add a Close() function that does that.

achow101 commented:
I've redone the refactor to keep `BerkeleyBatch` and introduce `DatabaseBatch` as an abstract class. Database transactions, cursors, and modifications occur via `BerkeleyBatch`.

This change also results in changes to the `Flush` and `Close` logic. I've decided to keep the `shutdown` argument in `BerkeleyEnvironment::Flush`. So `BerkeleyDatabase::Flush` calls that with `shutdown == false` and `BerkeleyDatabase::Close` calls it with `shutdown == true`. This makes the changes there far fewer and should be easier to review.

The commits have also been reordered so that moves occur first so this is less painful to modify later.

MarcoFalke commented:
Can the move-only stuff be split up in its own pr? That seems like something a lot more people could review than the bdb-refactors.

achow101 commented:
I've started breaking this down into separate PRs. Some of the simpler stuff is moved up to the front so they can be merged first.

* Mostly simple moveonly things: #19290
  * walletdb: Make SpliWalletFilePath non-static
  * walletdb: Add IsBDBWalletLoaded to look for BDB wallets specifically
  * walletdb: move IsWalletLoaded to walletdb.cpp
  * walletdb: moveonly: Move BerkeleyBatch Cursor and Txn funcs to cpp
  * walletdb: Move BDB specific things into bdb.{cpp/h}
* Refactor Read, Write, Erase, Exists: #19292
  * walletdb: refactor Read, Write, Erase, and Exists into non-template func

ryanofsky commented on src/wallet/bdb.cpp:
```diff
@@ -528,17 +528,17 @@ bool BerkeleyBatch::Rewrite(BerkeleyDatabase& database, const char* pszSkip)
 
-                    Dbc* pcursor = db.GetCursor();
-                    if (pcursor)
+                    if (db.CreateCursor())
```
In commit "walletdb: Handle cursor internally" (84bc5c1705691f4ecdfb8bcc90eb1bc4970281dc)

As noted https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437625076, the error handling for failing to get a cursor here is broken. It had been broken before this commit, but now this commit is introducing a new way for getting a cursor to fail (the case where m_cursor is not null), so this might be making the problem worse. A simple way to fix this while keeping status quo would be to assert(!m_cursor) in CreateCursor instead of returning false. Otherwise it probably would be good to do improve the error handling here and at least log or return failure.

achow101 replied:
I've changed it to assert.

ryanofsky commented on src/wallet/bdb.h:
```diff
@@ -217,6 +217,7 @@ class BerkeleyBatch
     Db* pdb;
     std::string strFile;
     DbTxn* activeTxn;
+    Dbc* m_cursor;
```
In commit "walletdb: Handle cursor internally" (84bc5c1705691f4ecdfb8bcc90eb1bc4970281dc)

Could the commit message for this say something about the motivation for this change? It seems good, but I'm also not sure I understand why it's happening.

achow101 replied:
Done.

This is happening because `Dbc` is BDB specific so a generic `DatabaseBatch` class can't return them.

ryanofsky commented on src/wallet/walletdb.cpp:
```diff
@@ -880,8 +886,8 @@ DBErrors WalletBatch::FindWalletTx(std::vector<uint256>& vTxHash, std::list<CWal
         }
-        pcursor->close();
     } catch (...) {
+        m_batch.CloseCursor();
```
In commit "walletdb: Handle cursor internally" (84bc5c1705691f4ecdfb8bcc90eb1bc4970281dc)

Is it intentional to close the cursor in the catch block and stop closing the cursor in the try block? It looks like it could be an accidental change

achow101 replied:
Yes, this was intentional. I believe previously we could have had a memory leak as the cursor could possible never get closed.

Before the while loop is exited, we always close the cursor. So having this at the end of the try block isn't useful. However it is also possible that the while loop is exited by an exception, in which case the cursor would never get closed as the catch block doesn't close it, and the cursor wasn't being closed afterwards. Moving the `CloseCursor` into the catch block handles this case.

ryanofsky commented on src/wallet/bdb.h:
```diff
@@ -303,8 +304,9 @@ class BerkeleyBatch
-    int ReadAtCursor(Dbc* pcursor, CDataStream& ssKey, CDataStream& ssValue);
+    bool CreateCursor();
+    bool ReadAtCursor(CDataStream& ssKey, CDataStream& ssValue, bool& complete);
+    void CloseCursor();
```
In commit "walletdb: Handle cursor internally" (84bc5c1705691f4ecdfb8bcc90eb1bc4970281dc)

Would it make sense to call CloseCursor in the BerkeleyBatch destructor? It seems like it could guard against memory leaks, and maybe avoid the need to add CloseCursor calls some of the places where they are added.

achow101 replied:
Done (I put it in `Close` which is called by the desctructor).

ryanofsky reviewed (commented):
Start rereviewing (will update list below with progress).
- [X] cc05c351c0e4a62eed81996b5370e64b28b985fe walletdb: Handle cursor internally (1/15)
- [ ] 11f7966982364911f4ee88df787961b110d24a38 Add Create*WalletDatabase functions (2/15)
- [ ] 2dd5707793a448b6425490e17847d4b12cb53020 scripted-diff: Replace WalletDatabase::Create* with CreateWalletDatabase (3/15)
- [ ] 37c656b671ba1e17df5f8cfb2ae7776555e61313 Remove WalletDatabase::Create, CreateMock, and CreateDummy (4/15)
- [ ] 622b5cbddfab6496f4334271a8aedeaff3a6999d walletdb: Combine VerifyDatabaseFile and VerifyEnvironment (5/15)
- [ ] 73888f0bf833d28d951a3cf597c772c7a3d87228 walletdb: Move PeriodicFlush into WalletDatabase (6/15)
- [ ] 8c903f535019b4820cd2259eb1d77033d32b066a walletdb: Move Db->open to BerkeleyDatabase::Open (7/15)
- [ ] b0e0ecca36728c79e985acfbafc6737d5d87570f walletdb: Use a global g_fileids instead of m_fileids for each env (8/15)
- [ ] 54ee8905d23213d6165fe45601627d2aadee1bb7 walletdb: Move BerkeleyDatabase::Flush(true) to Close() (9/15)
- [ ] c5b005f34ccc0d0a40486da37394a3556e1fa14b walletdb: Add BerkeleyDatabase::RemoveRef (10/15)
- [ ] baca4d0769b63578a8e4a94beb01f8f750c1fb6f walletdb: track database file use as m_refcount within BerkeleyDatabase (11/15)
- [ ] 029b13023dd3b695bcbf65a8fa4e89640b86ddd9 walletdb: Move Rewrite into BerkeleyDatabase (12/15)
- [ ] 58fc08efd4641cdeec74e9af1dd71a12fd09efbc walletdb: Remove BerkeleyBatch friend class from BerkeleyDatabase (13/15)
- [ ] b11e2ad21ed264307e746fb198c5ddeb8f23c954 walletdb: Add MakeBatch function to BerkeleyDatabase and use it (14/15)
- [ ] d850984306f8b605ae5a35dd19018b734f2bbd55 walletdb: Introduce WalletDatabase and DatabaseBatch abstract classes (15/15)

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -103,6 +122,8 @@ void BerkeleyEnvironment::Close()
             database.m_db->close(0);
             database.m_db.reset();
+            LOCK(cs_db);
+            g_fileids.erase(database.m_file_path);
```
Ah, I see how that could theoretically happen. I did a bit of testing and wasn't able to trigger a spurious duplicate fileid error. I believe that's because closing database A also closed database B since nothing was using it. So to get this error, B needs to be in use.

The reason this was failing tests was because I corrected a mistake in the patch. You have `g_fileids.erase(m_file_path);` which should actually be `erased = g_fileids.erase(m_file_path);`. Otherwise the following `assert(erased == 1);` doesn't do anything. But because `BerkeleyEnvironment::Close` removes the fileid before the destructor is called, this assertion will fail.

So instead, I've dropped that assertion but still kept the `g_fileids.erase(m_file_path);`.

achow101 committed:
walletdb: Refactor DatabaseBatch abstract class from BerkeleyBatch

Sjors commented:
I get similar errors as [Travis](https://travis-ci.org/github/bitcoin/bitcoin/jobs/699909476) on macOS (`./configure --enable-debug --with-incompatible-bdb --enable-werror`):
```
./wallet/bdb.h:167:17: error: non-static data member 'm_file_path' of 'BerkeleyDatabase' shadows member inherited from type 'WalletDatabase' [-Werror,-Wshadow-field]
1687    std::string m_file_path;
1688                ^
1689./wallet/db.h:166:17: note: declared here
1690    std::string m_file_path;
1691                ^
```

And:
```
./wallet/walletdb.h:208:17: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
        m_batch(std::move(database.MakeBatch(pszMode, _fFlushOnClose))),
                ^
```

achow101 commented:
> I get similar errors as Travis on macOS

Should be fixed now.

achow101 committed:
walletdb: Introduce AddRef and RemoveRef functions

Refactor mapFileUseCount increment and decrement to separate functions
AddRef and RemoveRef

achow101 committed:
walletdb: Add BerkeleyDatabase::Open dummy function

Adds an Open function for the class abstraction that does nothing for
now.

achow101 commented:
This PR is now superseded by the PRs listed in the OP. Even though the final one, #19335, is the same diff and commits as this one, I wanted to have it be separate without the history from this PR as it is conceptually unrelated to this refactor. The refactoring into the abstract classes occurs in #19334.
